File: backend/python/pcb_to_netlist/6_map_component_names.py
# file: 6_map_component_names.py
# Replace component IDs with designators in nets

import sys # argv, exit

def main():
    if len(sys.argv) != 4: sys.exit(1)
    pcb, nets_f, out_f = sys.argv[1:4]
    cmap = {}
    with open(pcb, 'r') as f:
        for l in f:
            if not l.startswith('RECORD=Component'): continue
            p = {x.split('=')[0]: x.split('=')[1] for x in l.split('|') if '=' in x}
            if 'ID' in p and 'SOURCEDESIGNATOR' in p: cmap[p['ID']] = p['SOURCEDESIGNATOR']
    with open(nets_f, 'r') as f_in, open(out_f, 'w') as f_out:
        for l in f_in:
            if not l.strip() or ':' not in l: continue
            nid, conns = l.strip().split(':', 1)
            mapped = []
            for c in conns.split(','):
                if '_' not in c: continue
                cid, pin = c.split('_', 1)
                mapped.append(f"{cmap.get(cid, cid)}_{pin}")
            f_out.write(f"{nid}:{','.join(mapped)}\n")

if __name__ == "__main__":
    main()

File: src/typescript/pcb_to_netlist/6_map_component_names.ts
import fs from 'fs';

export function mapComponentNames(pcbPath: string, netsPath: string, outPath: string): void {
    const cmap: Record<string, string> = {};
    const pcbContent = fs.readFileSync(pcbPath, 'utf8');
    
    for (const line of pcbContent.split(/\r?\n/)) {
        if (!line.startsWith('RECORD=Component')) continue;
        const p: Record<string, string> = {};
        line.split('|').forEach(x => {
            if (x.includes('=')) {
                const [k, v] = x.split('=', 2);
                p[k] = v;
            }
        });
        if (p['ID'] && p['SOURCEDESIGNATOR']) {
            cmap[p['ID']] = p['SOURCEDESIGNATOR'];
        }
    }
    
    const netsContent = fs.readFileSync(netsPath, 'utf8');
    const output: string[] = [];
    
    for (const line of netsContent.split(/\r?\n/)) {
        if (!line.trim() || !line.includes(':')) continue;
        const [nid, conns] = line.split(':', 2);
        const mapped = conns.split(',').map(c => {
            if (!c.includes('_')) return c;
            const [cid, pin] = c.split(/_(.+)/);
            return `${cmap[cid] || cid}_${pin}`;
        });
        output.push(`${nid}:${mapped.join(',')}`);
    }
    
    fs.writeFileSync(outPath, output.join('\n'));
}
